src/ast/irbuilderbpf.cpp

```diff
@@ -543,6 +543,64 @@ void IRBuilderBPF::CreateForEachMapElem(Value *ctx,
   CreateHelperErrorCond(ctx, call, libbpf::BPF_FUNC_for_each_map_elem, loc);
 }
 
+void IRBuilderBPF::CreateCheckSetRecursion(Value *ctx, const location &loc)
+{
+  assert(ctx && ctx->getType() == GET_PTR_TY());
+
+  const std::string map_ident = to_string(MapType::RecursionPrevention);
+
+  AllocaInst *key = CreateAllocaBPF(getInt32Ty(), "lookup_key");
+  CreateStore(getInt32(0), key);
+
+  auto type = CreateUInt64();
+  Value *value = CreateMapLookupElem(ctx, map_ident, key, type, loc);
+
+  Function *parent = GetInsertBlock()->getParent();
+  BasicBlock *value_is_set_block = BasicBlock::Create(module_.getContext(),
+                                                      "value_is_set",
+                                                      parent);
+  BasicBlock *value_is_unset_block = BasicBlock::Create(module_.getContext(),
+                                                        "value_is_unset",
+                                                        parent);
+  Value *condition = CreateICmpEQ(value, getInt64(1), "value_set_condition");
+  CreateCondBr(condition, value_is_set_block, value_is_unset_block);
+
+  SetInsertPoint(value_is_set_block);
+
+  CreateLifetimeEnd(key);
+
+  /*
+   * The counter is set, we need to exit early from the probe.
+   * This should only happend for the functions that can lead
+   * to a recursive crash e.g. "queued_spin_lock_slowpath"
+   */
+  CreateRet(getInt64(0));
+
+  SetInsertPoint(value_is_unset_block);
+
+  AllocaInst *set_value = CreateAllocaBPF(getInt64Ty(), "set_value");
+  CreateStore(getInt64(1), set_value);
+  CreateMapUpdateElem(ctx, map_ident, key, set_value, loc);
```

Comment:
Pls remove. Always update map value in place. No need for this overhead.
Also since it's filtering for kfuncs attached to spin_locks there is no guarantee that irqs are disabled,
so it's better to do __sync_fetch_and_add() instead of vanilla increment.
Only if kernel is really old fall back to ldx, +1, stx.

Reply:
Good catch. I did this for the other per-cpu maps in bpftrace but forgot to do it here lol.

> Also since it's filtering for kfuncs attached to spin_locks there is no guarantee that irqs are disabled,
so it's better to do __sync_fetch_and_add() instead of vanilla increment.

What if we are just less strict about the comparison and just check if it's not 0, then even if there is a race condition and it ends up being 2 instead of 1 then it still counts as set.

Reply:
Not doing atomic means that a cpu will read zero, assume its fine, then irq, then the same prog reads zero again and both will execute.
That's why in the kernel we do:

Reply:
Instead of doing an addition, I'm just doing an exchange because we just care if the number is 1 or 0 and I think this should work regardless of kernel version (hopefully).
```
CREATE_ATOMIC_RMW(AtomicRMWInst::BinOp::Xchg,
                    cast,
                    getInt64(1),
                    8,
                    AtomicOrdering::SequentiallyConsistent);
```